File: RustNet/dotnet/src/RustNet.Buses/Can.cs
using RustNet.Core;

namespace RustNet.Buses;

/// <summary>One received CAN frame.</summary>
public class CanFrame
{
    public int Id;
    public bool Extended;
    public bool Rtr;
    public byte[] Data = new byte[0];
}

/// <summary>
/// Classic CAN 2.0 bus master. Frames are queued in the controller FIFO;
/// <see cref="Read"/> returns null when nothing is pending.
/// </summary>
public static class Can
{
    [InternalCall]
    public static void Init(int bus, int bitrate, bool loopback) => throw new RuntimeOnlyException();

    [InternalCall]
    public static void Write(int bus, int id, byte[] data) => throw new RuntimeOnlyException();

    [InternalCall]
    public static int Available(int bus) => throw new RuntimeOnlyException();

    /// <summary>Raw packed frame: id u32 LE | flags | len | data. Prefer <see cref="Read"/>.</summary>
    [InternalCall]
    public static byte[] ReadRaw(int bus) => throw new RuntimeOnlyException();

    /// <summary>Hardware acceptance filter: accept when (frameId &amp; mask) == (id &amp; mask).</summary>
    [InternalCall]
    public static void SetFilter(int bus, int id, int mask) => throw new RuntimeOnlyException();

    /// <summary>Pop the next received frame, or null when the FIFO is empty.</summary>
    public static CanFrame Read(int bus)
    {
        byte[] raw = ReadRaw(bus);
        if (raw == null)
        {
            return null;
        }
        CanFrame f = new CanFrame();
        f.Id = raw[0] | (raw[1] << 8) | (raw[2] << 16) | (raw[3] << 24);
        f.Extended = (raw[4] & 1) != 0;
        f.Rtr = (raw[4] & 2) != 0;
        int len = raw[5];
        byte[] data = new byte[len];
        for (int i = 0; i < len; i++)
        {
            data[i] = raw[6 + i];
        }
        f.Data = data;
        return f;
    }
}

File: RustNet/runtime/rustnet-hal/src/can.rs

#[allow(unused_imports)]
use alloc::{boxed::Box, string::{String, ToString}, vec::Vec};
use crate::HalResult;

/// A classic CAN 2.0 frame (11-bit or 29-bit identifier, 0..=8 data bytes).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CanFrame {
    pub id: u32,
    /// 29-bit extended identifier when true, 11-bit standard otherwise.
    pub extended: bool,
    /// Remote transmission request (no data).
    pub rtr: bool,
    pub data: Vec<u8>,
}

impl CanFrame {
    pub fn new(id: u32, data: &[u8]) -> Self {
        CanFrame { id, extended: id > 0x7FF, rtr: false, data: data.to_vec() }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct CanConfig {
    pub bitrate: u32,
    /// Frames transmitted are also received locally (self-test mode).
    pub loopback: bool,
    /// Listen-only: no ACKs are generated on the wire.
    pub silent: bool,
}

impl Default for CanConfig {
    fn default() -> Self {
        CanConfig { bitrate: 500_000, loopback: false, silent: false }
    }
}

pub trait CanBus: Send {
    fn configure(&mut self, config: CanConfig) -> HalResult<()>;
    /// Queue a frame for transmission. Non-blocking; the controller FIFO
    /// drains in the background.
    fn transmit(&mut self, frame: &CanFrame) -> HalResult<()>;
    /// Pop the next received frame, if any (non-blocking).
    fn receive(&mut self) -> HalResult<Option<CanFrame>>;
    /// Number of frames waiting in the receive FIFO.
    fn rx_pending(&self) -> usize;
    /// Hardware acceptance filter: a frame is accepted when
    /// `(frame.id & mask) == (id & mask)`. Pass mask 0 to accept all.
    fn set_filter(&mut self, id: u32, mask: u32) -> HalResult<()>;
}
